pkg/broker/work_engine.go

```diff
@@ -7,32 +7,59 @@ type Work interface {
 	Run(token string, msgBuffer chan<- WorkMsg)
 }
 
+// WorkTopic - Topic jobs can publish messages to, and subscribers can listen to
+type WorkTopic string
+
 // WorkEngine - a new engine for doing work.
 type WorkEngine struct {
-	msgBuffer chan WorkMsg
+	topics map[WorkTopic]chan WorkMsg
 }
 
 // NewWorkEngine - creates a new work engine
 func NewWorkEngine(bufferSize int) *WorkEngine {
-	return &WorkEngine{
-		msgBuffer: make(chan WorkMsg, bufferSize),
-	}
+	return &WorkEngine{topics: make(map[WorkTopic]chan WorkMsg)}
 }
 
-// StartNewJob - Starts a job in an new goroutine. returns token, or generated token if an empty token is passed in.
-func (engine *WorkEngine) StartNewJob(token string, work Work) string {
+//msgBuffer: make(chan WorkMsg, bufferSize),
+
+// StartNewJobOnTopic - Starts a job in an new goroutine, reporting to a specific topic.
+// returns token, or generated token if an empty token is passed in.
+func (engine *WorkEngine) StartNewJobOnTopic(
+	token string, work Work, topic WorkTopic,
+) string {
 	var jobToken string
 
 	if token == "" {
 		jobToken = uuid.New()
 	} else {
 		jobToken = token
 	}
-	go work.Run(jobToken, engine.msgBuffer)
+
+	msgBuffer, topicExists := engine.topics[topic]
+	if !topicExists {
+		msgBuffer = make(chan WorkMsg)
+		engine.topics[topic] = msgBuffer
+	}
+
+	go work.Run(jobToken, msgBuffer)
 	return jobToken
 }
 
-// AttachSubscriber - Attach a subscriber to the engine. Will send the WorkMsg to the subscribers through the message buffer.
-func (engine *WorkEngine) AttachSubscriber(subscriber WorkSubscriber) {
-	subscriber.Subscribe(engine.msgBuffer)
+// AttachSubscriberToTopic - Attach a subscriber a specific messaging topic.
+// Will send the WorkMsg to the subscribers through the message buffer.
+func (engine *WorkEngine) AttachSubscriberToTopic(
+	subscriber WorkSubscriber, topic WorkTopic,
+) {
+	msgBuffer, topicExists := engine.topics[topic]
+	if !topicExists {
+		msgBuffer = make(chan WorkMsg)
+		engine.topics[topic] = msgBuffer
+	}
```

Comment:
At least we need to check that the topic is in the enum IMO. I can see the value in lazy initialization here but if we are using an enum to make sure only certain work topics are allowed then we need to check it because 
```go
WorkTopic("somethingNewTopic")
```
is a valid work topic to the compiler.